CLUSTER.md

```diff
@@ -0,0 +1,450 @@
+    - [Publish/Subscribe](#publish/subscribe) 
+    - [Reference Implementation](#reference-implementation)
+        - [Road to Production](#road-to-production) 
+        - [Potential Enhancements](#potential-enhancements)
+<!-- /toc -->
+
+## Summary
+
+A new approach to Redis Clustering is proposed in order to:
+
+- Provide predictable behavior & correctness guarantees across the various possible failure modes
+- Create a minimal, well-defined internal interface to the clustering code so alternative approaches can easily replace the
+  entire clustering functionality at compile time.
+- Support a significantly higher number of shards
+- Significantly reduce failover time to under 10 seconds (at worse) and under 5 seconds in some configurations 
+- Allow for alternative consistency mode of operation
+- Long term: Support a clustered, but un-sharded deployment mode with no cross-slot limitations, which will allow us to deprecate all Sentinel code
+- Long term: Enable centralized management for global configuration settings & Redis Functions
+
+This new approach is:
+- Based on a two-tier strongly consistent, consensus-based control plane
+- Maintains full backwards compatibility with data path Cluster commands (MOVED, NODES, SLOTS, ...), but not 
+  administrative commands (SETSLOT, FORGET, MEET,...)
+- Has a consistent source of truth for all cluster metadata
+- Requires the shards to reconcile their state according to the above source
+- Maintains the ability to scale down to a bare minimum 3 redis-server processes
+- Is called Flotilla :)
+
+And is intended to fully replace the current clustering implementation after a deprecation period in which the two 
+implementations will live side-by-side with compile time flag to chose between the two.
+
+## Motivation
+
+The current Redis Cluster is based on Gossip with the use of epochs to attempt to 
+control/reduce the impact of concurrent changes. It is easy to set up but has a lot of issues when performing day 2
+operations & scaling.
+
+For example:
+- Reliance on primaries (that may fail-over and move around) for quorum makes it impossible to guarantee availability using rack awareness.
+- Node removal has an inherent race condition which can lead to nodes being re-added
+- No single source of truth about who the primary is for a given slot range
+
+Our main goal with this new approach is to provide a more robust solution, with an easy-to-understand
+behavior and predictable failure modes. Beyond this, we also aim to substantially improve the scalability in order to
+be able to easily support a cluster with thousands of shards.
+
+## Design
+
+### Terminology
+
+| Term | Description |
+| --- | --- |
+| Node | A Single redis-server process |
+| Shard | A logical entity comprised of a group of nodes owning the same slots. It is automatically allocated when a new unique slot range is assigned to a node and automatically deallocated when all slots are moved away |
```

Comment:
The reason I wanted to avoid it is to reduce the admin overhead -- but you're right that some use cases will require it.
Config option or maybe have it default to auto-mangement and switch to admin managed once an admin calls the new CREATE_SHARD API?
(I don't think we need to support a hybrid mode where one can create and delete shards both automatically and manually -- it would be very error prone)

Reply:
Another option would be to allow setting shard ID in the NODE_JOIN API (and still have the deletion happen automatically)

Reply:
I've added the option for setting IDs on entity creation.